crates/lsp/src/tower_lsp.rs

```diff
@@ -65,31 +66,97 @@ pub(crate) enum LspRequest {
 }
 
 #[allow(clippy::large_enum_variant)]
-#[derive(Debug)]
+#[derive(Debug, IntoStaticStr)]
 pub(crate) enum LspResponse {
     Initialize(InitializeResult),
     DocumentFormatting(Option<Vec<TextEdit>>),
     Shutdown(()),
     AirViewFile(String),
 }
 
+impl std::fmt::Display for LspNotification {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.into())
+    }
+}
+impl std::fmt::Display for LspRequest {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.into())
+    }
+}
+impl std::fmt::Display for LspResponse {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.into())
+    }
+}
+
+impl LspNotification {
+    fn trace(&self) -> TraceLspNotification {
+        TraceLspNotification { inner: self }
+    }
+}
+impl LspRequest {
+    fn trace(&self) -> TraceLspRequest {
+        TraceLspRequest { inner: self }
+    }
+}
+impl LspResponse {
+    fn trace(&self) -> TraceLspResponse {
+        TraceLspResponse { inner: self }
+    }
+}
+
+struct TraceLspNotification<'a> {
+    inner: &'a LspNotification,
+}
+struct TraceLspRequest<'a> {
+    inner: &'a LspRequest,
+}
+struct TraceLspResponse<'a> {
+    inner: &'a LspResponse,
+}
+
+impl std::fmt::Debug for TraceLspNotification<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.inner {
+            LspNotification::DidOpenTextDocument(params) => {
+                // Ignore the document itself in trace logs
+                f.debug_tuple(self.inner.into())
+                    .field(&params.text_document.uri)
+                    .field(&params.text_document.version)
+                    .field(&params.text_document.language_id)
+                    .finish()
+            }
+            _ => std::fmt::Debug::fmt(self.inner, f),
+        }
+    }
+}
```

Comment:
It ended up being easier to write a wrapper type that overrides specific variants but otherwise pushes through to the "normal" derived `Debug` implementation. If we try and implement `Debug` on `LspNotification` directly then we lose everything it auto-derives for us, which is actually quite a lot of repetitive code.

Reply:
I was thinking about this problem after our call yesterday. That's a nice solution.